File: cpp/lib_ecw.cpp
/**
 * @Time    :2025/09/24 23:55:26
 * @Author  :小 y 同 学
 * @公众号   :小y只会写bug
 * @CSDN    :https://blog.csdn.net/weixin_64989228?type=blog
 */

#include "lib_ecw.h"

/**
 * 梯形积分，注意x和y的位置
 */
double trapz(std::vector<double> y, std::vector<double> x)
{
    double sum_area = 0.0; //初始面积为0
    for (size_t i = 1; i < y.size(); ++i)
    {
        //计算梯形面积并累加
        sum_area += 0.5 * (y[i] + y[i - 1]) * (x[i] - x[i - 1]);
    }
    return sum_area;
}

/**
 * 读取srf数据，返回wave：单位nm，srf：0-1之间
 */
void read_srf_data(std::string filepath, std::vector<double> &wave, std::vector<double> &srf)
{
    std::ifstream infile(filepath);
    std::string line;
    double w, s;
    // =====> 忽略前面四行 <=====
    for (int i = 0; i < 4; ++i)
    {
        std::getline(infile, line);
    }

    // =====> 读取数据 到 wave 和 srf 数组<=====
    while (std::getline(infile, line))
    {
        std::istringstream iss(line);
        if (!(iss >> w >> s))
        {
            std::cout << "Error reading line: " << line << std::endl;
            break;
        }
        w = 1.0 / w * 1e7; // cm-1 to nm
        // std::cout << std::fixed << w << " " << s << std::endl;
        wave.push_back(w);
        srf.push_back(s);
    }
}

/**
 * 计算ecw
 */
double cal_ecw(std::vector<double> &wave, std::vector<double> &srf)
{
    // =====> 计算总面积和1/2面积 <=====
    double sum_srf = trapz(srf, wave);
    double sum_srf_2 = sum_srf / 2.0;

    // =====> 二分法 迭代 <=====
    double tmp_srf = -1;                     //初始面积
    double lwave = wave[0];                  //左区间
    double rwave = wave[wave.size() - 1];    //右区间
    double tmp_wave = (lwave + rwave) / 2.0; //中间点
    while (true)
    {
        // =====> 计算当前面积 <=====
        std::vector<double> srf_tmp;
        for (size_t i = 0; i < wave.size(); ++i)
        {
            if (wave[i] <= tmp_wave)
            {
                srf_tmp.push_back(srf[i]);
            }
            else
            {
                srf_tmp.push_back(0.0);
            }
        }
        tmp_srf = trapz(srf_tmp, wave);

        // =====> 更新左右区间 <=====
        if (std::abs(rwave - lwave) < 1e-6)
        {
            break;
        }
        else if (tmp_srf > sum_srf_2)
        {
            rwave = tmp_wave;
        }
        else
        {
            lwave = tmp_wave;
        }
        tmp_wave = (lwave + rwave) / 2.0;
    }
    return tmp_wave;
}

File: python/lib_ecw.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# @Time    :2025/09/22 20:36:15
# @Author  :小 y 同 学
# @公众号   :小y只会写bug
# @CSDN    :https://blog.csdn.net/weixin_64989228?type=blog
# @Discript:读取光谱响应函数，计算等效中心波长

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import os
import re

# =====>2025/09/22 21:51:48 设置绘图属性 <=====
# 标题
mpl.rcParams["axes.titlesize"] = 20
# x轴标签
mpl.rcParams["axes.labelsize"] = 16
# 刻度标签
mpl.rcParams["xtick.labelsize"] = 14
mpl.rcParams["ytick.labelsize"] = 14
# 图例
mpl.rcParams["legend.fontsize"] = 14
# 支持中文
config = {
    "font.family": "serif",  # 设置衬线字体
    "font.serif": ["SimSun"],  # 设置中文字体,SimSun为宋体
    "mathtext.fontset": "stix",  # 设置数学字体，用作英文字体
    "axes.unicode_minus": False,
}
mpl.rcParams.update(config)


# =====>2025/09/22 20:49:27 从srf文件中读取光谱响应函数 <=====
def read_srf_file(file_path):
    data = np.loadtxt(file_path, skiprows=4)  # 跳过前4行
    wave = data[:, 0]
    wave_nm = 1 / wave * 1e7  # 将波长单位从 cm-1 转 nm
    srf_data = data[:, 1]
    return wave_nm, srf_data


# =====>2025/09/22 20:49:08 绘制光谱响应函数图像 <=====
def plot_spectrum(
    wave, srf, title="Spectrum", ecw=None, lwave=None, rwave=None, save_path=None
):
    xlabel = "Wavelength (nm)"
    ylabel = "srf"

    # =====> 绘图 <=====
    plt.figure(figsize=(8, 6))
    plt.plot(wave, srf)
    plt.title(title)
    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    plt.grid()

    # =====> 绘制竖线 <=====
    if ecw:
        plt.axvline(x=ecw, color="r", linestyle="--", label=f"ecw: {ecw:.2f} nm")
    if lwave:
        plt.axvline(x=lwave, color="g", linestyle="-", label=f"Lwave: {lwave:.2f} nm")
    if rwave:
        plt.axvline(x=rwave, color="b", linestyle="-", label=f"Rwave: {rwave:.2f} nm")
    plt.legend(loc="upper right")

    # =====> 保存图片 <=====
    if save_path:
        if not os.path.exists(os.path.dirname(save_path)):
            os.makedirs(os.path.dirname(save_path))
        plt.savefig(save_path, dpi=600, bbox_inches="tight")
    else:
        plt.show()
    plt.close()


# =====>2025/09/22 21:06:15 计算等效中心频率 <=====
def cal_ecw(wave, srf):
    # 计算全部积分 面积
    sum_srf = np.trapz(srf, x=wave)
    sum_srf_2 = sum_srf / 2

    # 二分法 迭代求解积分中点
    tmp_srf = -1
    lwave = wave[0]
    rwave = wave[-1]
    tmp_wave = (lwave + rwave) / 2
    while True:
        # 计算当前频率下的积分
        tmp_srf = np.trapz(srf[wave <= tmp_wave], x=wave[wave <= tmp_wave])

        # ******************************
        # print("tmp_wave:", tmp_wave, " tmp_srf:", tmp_srf, " sum_srf_2:", sum_srf_2)
        # plot_spectrum(
        #     wave,
        #     srf,
        #     title="Spectrum",
        #     ecw=tmp_wave,
        #     lwave=lwave,
        #     rwave=rwave,
        #     # save_path=f"pic/ecw_{tmp_wave:.2f}.png",
        # )
        # ******************************

        # 判断是否达到目标积分
        if abs(rwave - lwave) < 1e-6:
            break
        # 如果当前积分小于目标积分，向右移动
        elif tmp_srf < sum_srf_2:
            lwave = tmp_wave
        # 否则向左移动
        else:
            rwave = tmp_wave
        tmp_wave = (lwave + rwave) / 2

    return tmp_wave
